File: examples/apps/python/msgpacketizer-py/src/packetizerpy/packetizer/subscriber.py
import logging
from typing import Callable, Dict

from .common import COBS_DELIMITER, Message, to_hex
from .decoder import decode


class Subscriber:
    def __init__(self):
        self._buffer = bytearray()
        self._callbacks: Dict[int, Callable[[Message], None]] = {}

    def feed(self, buffer: bytes):
        logging.debug(f"feed input: {to_hex(buffer)}")
        self._buffer += buffer
        msgs: list[Message] = []

        # find delimiter (0x00)
        while self._buffer.find(COBS_DELIMITER) != -1:
            # if delimiter found, split and process buffer one by one
            [chunk, self._buffer] = self._buffer.split(COBS_DELIMITER, maxsplit=1)
            logging.debug(f"buffer splited to chunk: {chunk}, rest: {self._buffer}")

            # decode packet
            decoded = decode(bytes(chunk))
            if decoded is not None:
                msgs.append(decoded)

        # iterate over received msgs and call callbacks
        for msg in msgs:
            logging.debug(f"received msg = index: {msg.index}, msg: {msg.data}")
            # do some stuff in user-defined callback function
            if msg.index in self._callbacks:
                self._callbacks[msg.index](msg)
            else:
                logging.warn(f"callback for index {msg.index} not found")

    def subscribe(self, index: int, callback: Callable[[Message], None]):
        self._callbacks[index] = callback

    def unsubscribe(self, index: int):
        del self._callbacks[index]

File: examples/apps/go/msgpacketizer-go/pkg/packetizer-go/subscriber.go
package packetizer

import (
	"bytes"

	"github.com/sirupsen/logrus"
)

type Subscriber struct {
	Buffer    []byte
	Msgs      []Message
	Callbacks map[uint8]func(Message)
}

func NewSubscriber() *Subscriber {
	return &Subscriber{
		Buffer:    make([]byte, 0),
		Msgs:      make([]Message, 0),
		Callbacks: map[uint8]func(Message){},
	}
}

func (s *Subscriber) Feed(buffer []byte) {
	logrus.Debugf("feed input: %s", ToHex(buffer))

	if len(buffer) == 0 {
		return
	}

	s.Buffer = append(s.Buffer, buffer...)

	for {
		// find delimiter
		i := bytes.IndexByte(s.Buffer, COBS_DELIMITER)
		if i < 0 {
			break
		}

		// decode packet
		chunk := s.Buffer[:i]
		s.Buffer = s.Buffer[i+1:]
		msg, err := Decode(chunk)
		if err != nil {
			logrus.Warnf("decode error: %s", err)
			continue
		}

		// append to msgs
		s.Msgs = append(s.Msgs, msg)
	}

	// iterate over received messages and call callbacks
	for _, msg := range s.Msgs {
		if callback, ok := s.Callbacks[msg.Index]; ok {
			callback(msg)
		}
	}

	// clear received messages
	s.Msgs = make([]Message, 0)
}

func (s *Subscriber) Subscribe(index uint8, callback func(Message)) {
	s.Callbacks[index] = callback
}

func (s *Subscriber) Unsubscribe(index uint8) {
	delete(s.Callbacks, index)
}
